build-common/src/org/jetbrains/kotlin/incremental/IncrementalJvmCache.kt

```diff
@@ -757,33 +757,35 @@ private class ConstantsClassVisitor : ClassVisitor(Opcodes.API_VERSION) {
 }
 
 private class InlineFunctionsClassVisitor(
-    private val inlineFunctionNames: Set<String>,
+    private val inlineFunctionSignatures: Set<String>,
     cv: ConstantsClassVisitor // Note: cv must not override the visitMethod (it will not be called with the current implementation below)
 ) : ClassVisitor(Opcodes.API_VERSION, cv) {
 
     private val result = LinkedHashMap<String, Long>()
-    private var dummyVersion: Int = -1
+    private var classVersion: Int? = null
 
-    override fun visit(version: Int, access: Int, name: String?, signature: String?, superName: String?, interfaces: Array<out String>?) {
+    override fun visit(version: Int, access: Int, name: String, signature: String?, superName: String?, interfaces: Array<out String>?) {
         super.visit(version, access, name, signature, superName, interfaces)
-        dummyVersion = version
+        classVersion = version
     }
 
     override fun visitMethod(access: Int, name: String, desc: String, signature: String?, exceptions: Array<out String>?): MethodVisitor? {
         if (access and Opcodes.ACC_PRIVATE == Opcodes.ACC_PRIVATE) return null
 
-        val dummyClassWriter = ClassWriter(0)
-        dummyClassWriter.visit(dummyVersion, 0, "dummy", null, AsmTypes.OBJECT_TYPE.internalName, null)
+        // Note: Here, functionSignature = name + descriptor.
+        // It is different from the `signature` parameter above, which is essentially a more detailed descriptor when generics are used
+        // (or null otherwise).
+        val functionSignature = JvmMemberSignature.Method(name, desc).asString()
+        if (functionSignature !in inlineFunctionSignatures) return null
 
-        return object : MethodVisitor(Opcodes.API_VERSION, dummyClassWriter.visitMethod(0, name, desc, null, exceptions)) {
-            override fun visitEnd() {
-                val jvmName = name + desc
-                if (jvmName !in inlineFunctionNames) return
+        val classWriter = ClassWriter(0)
 
-                val dummyBytes = dummyClassWriter.toByteArray()!!
+        // The `version` and `name` parameters are important (see KT-38857), the others can be null.
+        classWriter.visit(/* version */ classVersion!!, /* access */ 0, /* name */ "ClassWithOneMethod", null, null, null)
```

Comment:
Why did you remove "Object" as supertype?

Reply:
Nice catch, the `ClassNode()` approach in `JavaClassSnapshotter` has supertype = `null` so I do the same here for consistency.

Reply:
As long as ASM does not fail and it generates class that we can hash, it LGTM.